android/fakedapp/src/main/java/com/solana/mobilewalletadapter/fakedapp/MainViewModel.kt

```diff
@@ -20,19 +20,19 @@ import kotlinx.coroutines.flow.MutableStateFlow
 import kotlinx.coroutines.flow.asStateFlow
 import kotlinx.coroutines.flow.update
 import kotlinx.coroutines.sync.Mutex
-import kotlinx.coroutines.sync.Semaphore
 import kotlinx.coroutines.sync.withLock
 import java.io.IOException
 import java.util.concurrent.CancellationException
 import java.util.concurrent.ExecutionException
+import java.util.concurrent.TimeUnit
 import java.util.concurrent.TimeoutException
 import kotlin.random.Random
 
 class MainViewModel(application: Application) : AndroidViewModel(application) {
     private val _uiState = MutableStateFlow(UiState())
     val uiState = _uiState.asStateFlow()
 
-    private val mobileWalletAdapterClientMutex = Mutex()
+    private val mobileWalletAdapterClientMutex = Mutex() // allow only a single MWA connection at a time
```

Comment:
This isn't a standard `java.util.concurrent` primitive, but rather a Kotlin coroutines mutex:
`import kotlinx.coroutines.sync.Mutex`
It's functionally appropriate for use inside of coroutines.

Re: the 1-connection limit, that's not enforced by the library; you could connect to multiple wallets simultaneously, such as wallets on different hosts for some kind of multi-sig signing. This is a safety measure for the demo app, which isn't designed to handle multiple MWA client instances. We could just remove it, but it's enforcing a fakedapp design decision, and so I'd need to replace it with some other mutual exclusion scheme 🤷‍♂️

Reply:
Heh, I didn't know there was a special coroutine Mutex. However, it appears that they exist to protect shared resources. Each call in this class opens, executes an operation, then closes. What shared resource is the Mutex protecting? Or put another way, what is the Mutex protecting against - rapid subsequent library calls?

Again, I'm kind of coming at this from doing things the "Android" way. This isnt something most libraries do, even in sample code. Nothing about the operations in this class seem fundamentally more complex than what other io libraries are doing. I feel like something simpler can be done here.

Reply:
The resource is a logical resource - the right to perform wallet operations.

That said, happy to swap this out as well. Perhaps a channel paradigm for these requests, with a single consumer, that would integrate equally well with the suspend nature of the incoming calls.

My main thought here is that this doesn't belong in the library, because it's not a restriction imposed by the library - it's an app design decision on whether to allow multiple connections simultaneously.

Reply:
Since it's a synchronization change, I'll roll that into the Looper removal PR, to keep all the synchronization and threading changes in a single PR.

Reply:
Okay cool, that works for me! The PR is good to go.